jerry-ext/args/arg-transform-functions.c

```diff
@@ -0,0 +1,256 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+
+/**
+ * The common function to deal with optional arguments.
+ * The core transform function is provided by argument `func`.
+ *
+ * @return jerry undefined: the transformer passes,
+ *         jerry error: the transformer fails.
+ */
+jerry_value_t
+jerryx_arg_transform_optional (jerryx_arg_js_iterator_t *js_arg_iter_p, /**< available JS args */
+                               const jerryx_arg_t *c_arg_p, /**< native arg */
+                               jerryx_arg_transform_func_t func) /**< the core transform function */
+{
+  jerry_value_t js_arg = jerryx_arg_js_iterator_peek (js_arg_iter_p);
+
+  if (jerry_value_is_undefined (js_arg))
+  {
+    return js_arg;
+  }
+
+  return func (js_arg_iter_p, c_arg_p);
+} /* jerryx_arg_transform_optional */
+
+/**
+ * Tranform a JS argument to a double. Type coercion is not allowed.
+ */
+JERRYX_ARG_TRANSFORM_FUNC (jerryx_arg_transform_number_strict)
+{
+  jerry_value_t js_arg = jerryx_arg_js_iterator_pop (js_arg_iter_p);
+
+  if (!jerry_value_is_number (js_arg))
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE,
+                               (jerry_char_t *) "It is not a number.");
+  }
+
+  double *dest = c_arg_p->dest;
+  *dest = jerry_get_number_value (js_arg);
+
+  return jerry_create_undefined ();
+} /* JERRYX_ARG_TRANSFORM_FUNC */
+
+/**
+ * Tranform a JS argument to a double. Type coercion is allowed.
+ */
+JERRYX_ARG_TRANSFORM_FUNC (jerryx_arg_transform_number)
```

Comment:
To me, personally, it was a big help. We have cca. 30 functions with identical signature. This macro helped me a lot in ensuring that they are _really_ identical. But I can accept a verdict that function declarations should not be hidden by macros, too. It will be tedious to rewrite, but it will have to be done only once. Hopefully.